supervision/utils/file.py

```diff
@@ -1,10 +1,121 @@
+    bounding box coordinates and additional attributes like confidence, class ID, and tracker ID.
+
+    The class supports the capability to include custom data alongside the detection fields,
+    providing flexibility for logging various types of information.
+
+    Args:
+        filename (str): The name of the CSV file where the detections will be stored.
+                        Defaults to 'output.csv'.
+
+    Usage:
+        ```python
+        from supervision.utils.detections import Detections
+        # Initialize CSVSink with a filename
+        csv_sink = CSVSink('my_detections.csv')
+
+        # Assuming detections is an instance of Detections containing detection data
+        detections = Detections(...)
+
+        # Open the CSVSink context, append detection data, and close the file automatically
+        with csv_sink as sink:
+            sink.append(detections, custom_data={'frame': 1})
+        ```
+    """
+
+    def __init__(self, filename: str = "output.csv"):
+        self.filename = filename
+        self.file: Optional[open] = None
+        self.writer: Optional[csv.writer] = None
+        self.header_written = False
+        self.fieldnames = []  # To keep track of header names
+
+    def __enter__(self) -> "CSVSink":
+        self.open()
+        return self
+
+    def __exit__(
+        self,
+        exc_type: Optional[type],
+        exc_val: Optional[Exception],
+        exc_tb: Optional[Any],
+    ) -> None:
+        self.close()
+
+    def open(self) -> None:
+        self.file = open(self.filename, "w", newline="")
+        self.writer = csv.writer(self.file)
+
+    def close(self) -> None:
+        if self.file:
+            self.file.close()
+
+    def append(
+        self, detections: Detections, custom_data: Dict[str, Any] = None
+    ) -> None:
+        if not self.writer:
+            raise Exception(
+                f"Cannot append to CSV: The file '{self.filename}' is not open. Ensure that the 'open' method is called before appending data."
+            )
+        if not self.header_written:
+            self.write_header(detections, custom_data)
+        for i in range(len(detections.xyxy)):
+            self.write_detection_row(detections, i, custom_data)
+
+    def write_header(self, detections: Detections, custom_data: Dict[str, Any]) -> None:
+        base_header = [
+            "x_min",
+            "y_min",
+            "x_max",
+            "y_max",
+            "class_id",
+            "confidence",
+            "tracker_id",
+        ]
+        dynamic_header = sorted(
+            set(custom_data.keys()) | set(getattr(detections, "data", {}).keys())
+        )
+        self.fieldnames = base_header + dynamic_header
+        self.dynamic_fields = dynamic_header  # Store only the dynamic part
+        self.writer.writerow(self.fieldnames)
+        self.header_written = True
+
+    def write_detection_row(
```

Comment:
Let's separate the parsing logic from the writing logic. The parsing method should be static. This will allow us to write unit tests for parsing separately. 

Let me know if this explanation makes sense. I can give you more details if something needs to be clarified.